File: services/browser_monitor.py
"""
services/browser_monitor.py
------------------------------
Server-side counterpart to static/js/browser-monitor.js (Module 4,
Part 4).

The browser-side script detects tab switches, window blur/focus,
fullscreen exits, refresh attempts, copy/paste, right-click, and (best
effort) devtools usage, then POSTs each one to
routes/monitoring.py:/monitoring/browser-event. This module is the
single source of truth for which event_type slugs are valid and what
severity each one carries — routes/monitoring.py should never hardcode
that mapping itself, so a future module can add a new browser event
type (e.g. "screen_share_detected") by editing only this file.
"""

from services.event_logger import SEVERITY_INFO, SEVERITY_WARNING, SEVERITY_CRITICAL

# Every event_type the client is allowed to report, and the severity it
# carries. Anything not in this map is rejected by
# routes/monitoring.py — this is a small allowlist, not a place to
# trust arbitrary client input.
ALLOWED_BROWSER_EVENTS = {
    "tab_switch": SEVERITY_WARNING,
    "window_blur": SEVERITY_WARNING,
    "window_focus": SEVERITY_INFO,
    "fullscreen_exit": SEVERITY_CRITICAL,
    "page_refresh": SEVERITY_WARNING,
    "copy_attempt": SEVERITY_WARNING,
    "paste_attempt": SEVERITY_WARNING,
    "right_click": SEVERITY_WARNING,
    "devtools_detected": SEVERITY_CRITICAL,
}


class InvalidBrowserEventError(Exception):
    """Raised when the client reports an event_type outside the allowlist."""
    pass


def classify_event(event_type: str) -> str:
    """
    Return the severity for a given browser event_type.

    Raises InvalidBrowserEventError for anything not in the allowlist,
    so routes/monitoring.py can respond 400 Bad Request instead of
    silently logging arbitrary client-supplied event types.
    """
    if event_type not in ALLOWED_BROWSER_EVENTS:
        raise InvalidBrowserEventError(f"Unknown browser event type: '{event_type}'")
    return ALLOWED_BROWSER_EVENTS[event_type]

File: static/js/browser-monitor.js
/**
 * static/js/browser-monitor.js
 * -------------------------------
 * Module 4, Part 4: client-side browser integrity monitoring.
 *
 * Watches for tab switches, window blur/focus, fullscreen exit, page
 * refresh attempts, copy/paste, right-click, and (best effort)
 * developer-tools usage, and reports each one through a caller-supplied
 * `onEvent(eventType, description)` callback.
 *
 * Deliberately framework-agnostic and self-contained (no fetch calls
 * of its own) so it can be reused by any future monitoring page —
 * the page wires `onEvent` to whatever transport it wants (today:
 * POST /monitoring/browser-event, see monitoring.html).
 */

const ExamGuardBrowserMonitor = (function () {
    let active = false;
    let onEvent = function () {};
    let devtoolsCheckInterval = null;

    // Devtools detection is inherently best-effort in a browser sandbox
    // (there's no real API for "is devtools open"). This uses the
    // classic heuristic: measure the gap between outer and inner window
    // dimensions, which grows noticeably when devtools is docked.
    const DEVTOOLS_THRESHOLD_PX = 160;
    let devtoolsWasOpen = false;
    let fullscreenWasEntered = false;

    function handleVisibilityChange() {
        if (!active) return;
        if (document.hidden) {
            onEvent("tab_switch", "Candidate switched to another tab or minimized the window.");
        }
    }

    function handleWindowBlur() {
        if (!active) return;
        onEvent("window_blur", "Exam window lost focus.");
    }

    function handleWindowFocus() {
        if (!active) return;
        onEvent("window_focus", "Exam window regained focus.");
    }

    function handleFullscreenChange() {
    if (!active) return;

    const isFullscreen = !!(
        document.fullscreenElement ||
        document.webkitFullscreenElement
    );

    // Do not treat the initial state (before fullscreen was ever entered)
    // as an exit violation.
    if (!isFullscreen && fullscreenWasEntered) {
        onEvent(
            "fullscreen_exit",
            "Candidate exited fullscreen mode."
        );
    }

    if (isFullscreen) {
        fullscreenWasEntered = true;
    }
}

    function handleBeforeUnload(e) {
        if (!active) return;
        onEvent("page_refresh", "Candidate attempted to refresh or close the page.");
        // Standard "are you sure you want to leave" prompt — also acts
        // as a soft deterrent, not just a detector.
        e.preventDefault();
        e.returnValue = "";
    }

    function handleCopy() {
        if (!active) return;
        onEvent("copy_attempt", "Copy action attempted during the exam.");
    }

    function handlePaste() {
        if (!active) return;
        onEvent("paste_attempt", "Paste action attempted during the exam.");
    }

    function handleContextMenu(e) {
        if (!active) return;
        onEvent("right_click", "Right-click context menu attempted.");
        e.preventDefault();
    }

    function checkDevtools() {
        if (!active) return;
        const widthGap = window.outerWidth - window.innerWidth;
        const heightGap = window.outerHeight - window.innerHeight;
        const isOpen = widthGap > DEVTOOLS_THRESHOLD_PX || heightGap > DEVTOOLS_THRESHOLD_PX;

        if (isOpen && !devtoolsWasOpen) {
            onEvent("devtools_detected", "Browser developer tools usage detected.");
        }
        devtoolsWasOpen = isOpen;
    }

    function start(eventCallback) {
        if (active) return;
        active = true;
        onEvent = eventCallback || onEvent;
        devtoolsWasOpen = false;
        fullscreenWasEntered = false;

        document.addEventListener("visibilitychange", handleVisibilityChange);
        window.addEventListener("blur", handleWindowBlur);
        window.addEventListener("focus", handleWindowFocus);
        document.addEventListener("fullscreenchange", handleFullscreenChange);
        document.addEventListener("webkitfullscreenchange", handleFullscreenChange);
        window.addEventListener("beforeunload", handleBeforeUnload);
        document.addEventListener("copy", handleCopy);
        document.addEventListener("paste", handlePaste);
        document.addEventListener("contextmenu", handleContextMenu);

        devtoolsCheckInterval = setInterval(checkDevtools, 1500);
    }

    function stop() {
        if (!active) return;
        active = false;

        document.removeEventListener("visibilitychange", handleVisibilityChange);
        window.removeEventListener("blur", handleWindowBlur);
        window.removeEventListener("focus", handleWindowFocus);
        document.removeEventListener("fullscreenchange", handleFullscreenChange);
        document.removeEventListener("webkitfullscreenchange", handleFullscreenChange);
        window.removeEventListener("beforeunload", handleBeforeUnload);
        document.removeEventListener("copy", handleCopy);
        document.removeEventListener("paste", handlePaste);
        document.removeEventListener("contextmenu", handleContextMenu);

        if (devtoolsCheckInterval) {
            clearInterval(devtoolsCheckInterval);
            devtoolsCheckInterval = null;
        }
    }

    return { start: start, stop: stop };
})();
